core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/cluster/ccm/upgrade/UpgradeCcmService.java

```diff
@@ -98,21 +138,56 @@ public void updateTunnel(Long stackId) {
     }
 
-    public void pushSaltState(Long stackId) {
+    public void pushSaltState(Long stackId, Long clusterId) {
+        InMemoryStateStore.putStack(stackId, PollGroup.POLLABLE);
+        if (clusterId != null) {
+            InMemoryStateStore.putCluster(clusterId, PollGroup.POLLABLE);
+        }
+        clusterServiceRunner.redeployStates(stackId);
+        clusterServiceRunner.redeployGatewayPillar(stackId);
     }
 
-    public void reconfigureNginx(Long stackId) {
+    public void reconfigureNginx(Long stackId) throws CloudbreakOrchestratorException {
+        upgradeCcmOrchestratorService.reconfigureNginx(stackId);
     }
 
     public void registerClusterProxy(Long stackId) {
+        Optional<ConfigRegistrationResponse> configRegistrationResponse = clusterProxyService.reRegisterCluster(stackId);
+        configRegistrationResponse.ifPresentOrElse(c -> LOGGER.debug(c.toString()), () -> LOGGER.debug("No clusterproxy register response for {}", stackId));
     }
 
     public void healthCheck(Long stackId) {
-    }
-
-    public void removeAgent(Long stackId) {
-    }
-
-    public void deregisterAgent(Long stackId) {
+        Set<String> unhealthyHosts;
+        try {
+            unhealthyHosts = healthCheckService.getUnhealthyHosts(stackId);
+        } catch (RuntimeException ex) {
+            throw new CloudbreakServiceException("Cannot get host statuses, CM is likely not accessible. Need to roll back CCM upgrade to previous version.");
+        }
+        if (!unhealthyHosts.isEmpty()) {
+            LOGGER.info("There are unhealthy hosts after registering to Cluster Proxy: {}", unhealthyHosts);
+            throw new CloudbreakServiceException("One or more Gateway instance is not available. Need to roll back CCM upgrade to previous version.");
+        } else {
+            LOGGER.info("All hosts are healthy after registering to Cluster Proxy.");
+        }
+    }
+
+    public void removeAgent(Long stackId, Tunnel oldTunnel) throws CloudbreakOrchestratorException {
+        if (oldTunnel == Tunnel.CCM) {
+            upgradeCcmOrchestratorService.disableMina(stackId);
+        } else if (oldTunnel == Tunnel.CCMV2) {
+            upgradeCcmOrchestratorService.disableInvertingProxyAgent(stackId);
+        }
+    }
+
+    public void deregisterAgent(Long stackId, Tunnel oldTunnel) {
+        Stack stack = stackService.getById(stackId);
+        if (oldTunnel == Tunnel.CCM) {
+            String keyId = CcmResourceUtil.getKeyId(stack.getResourceCrn());
+            String userCrn = ThreadBasedUserCrnProvider.getUserCrn();
+            ccmResourceTerminationListener.deregisterCcmSshTunnelingKey(
+                    userCrn, Crn.safeFromString(stack.getResourceCrn()).getAccountId(), keyId, stack.getMinaSshdServiceId());
+        } else if (oldTunnel == Tunnel.CCMV2) {
+            ccmV2AgentTerminationListener.deregisterInvertingProxyAgent(stack.getCcmV2AgentCrn());
+        }
```

Comment:
please add a bit of a log here as well, thank you!

Reply:
ok